File: src/skill_xray/llm/privacy.py
"""Best-effort credential removal before constructing an opt-in LLM request."""

import re

from ..checks.supply_chain import _SECRET_RULES, _sanitize_source

_PEM = re.compile(
    r"-----BEGIN (?:[A-Z0-9]+ )*PRIVATE KEY(?: BLOCK)?-----.*?"
    r"(?:-----END (?:[A-Z0-9]+ )*PRIVATE KEY(?: BLOCK)?-----|\Z)", re.DOTALL)
_AUTH = re.compile(r"\b(?:Bearer|Basic)[ \t]+[A-Za-z0-9._~+/=-]+", re.IGNORECASE)
_VALUE_GAP = r"[ \t]*(?:\n(?:[ \t]*\n)*[ \t]+)?"
_NAMED = re.compile(
    r'''(?ims)((?<![\w-])(?=[\w-]*(?:token|password|passwd|secret|api[_-]?key|authorization))'''
    r'''[\w-]+'''
    r"""["']?[ \t]*[:=]""" + _VALUE_GAP + r")(?:[!&][^\s]*" + _VALUE_GAP + r"){0,2}("
    r'"""(?:\\.|(?!""").)*(?:"""|\Z)|'
    r"'''(?:\\.|(?!''').)*(?:'''|\Z)|"
    r'''"(?:\\.|[^"\\])*(?:"|\Z)|'''
    r"""'(?:''|\\.|[^'\\])*(?:'|\Z)|"""
    r"[^\n]+)")
_CONTINUATION = re.compile(r"\n([ \t]*)([^\n]*)")
_URL = re.compile(r'''(?<![a-z0-9+.-])[a-z][a-z0-9+.-]*://[^\s<>"']+''', re.IGNORECASE)


def redact(text):
    """Keep source lines stable; arbitrary or encoded secrets are not reliably recognizable."""
    text = _PEM.sub(lambda m: "[REDACTED]" + "\n" * m[0].count("\n"), text)
    for _, pattern, _ in _SECRET_RULES:
        text = pattern.sub("[REDACTED]", text)
    text = _AUTH.sub(lambda m: "[REDACTED]" + "\n" * m[0].count("\n"), text)
    parts, pos = [], 0
    while m := _NAMED.search(text, pos):
        prefix = text[text.rfind("\n", 0, m.start()) + 1:m.start()].rstrip("\"'")
        indent = len(prefix) if not prefix.strip(" \t-") else len(prefix) - len(prefix.lstrip())
        value_prefix = text[text.rfind("\n", 0, m.start(2)) + 1:m.start(2)]
        if "\n" in text[m.start():m.start(2)] and len(value_prefix) <= indent:
            parts.append(text[pos:m.start(2)])
            pos = m.start(2)
            continue
        end = m.end()
        if not m[2].startswith(('"', "'")):
            # A sibling YAML key is not a continuation of the credential value.
            while following := _CONTINUATION.match(text, end):
                if following[2] and len(following[1]) <= indent:
                    break
                end = following.end()
        parts.extend((text[pos:m.start()], m[1], "[REDACTED]",
                      "\n" * text[m.end(1):end].count("\n")))
        pos = end
    text = "".join(parts) + text[pos:]
    return _URL.sub(lambda m: _sanitize_source(m[0]), text)

File: internal/llm/privacy.go
package llm

import (
	"regexp"
	"slices"
	"strings"
	"unicode"
	"unicode/utf8"

	"github.com/dlclark/regexp2"

	"github.com/traceforce/skill-xray/internal/pytext"
	"github.com/traceforce/skill-xray/internal/supplychain"
)

// gap is _VALUE_GAP: the whitespace (with blank lines) between a key's separator and its value.
const gap = `[ \t]*(?:\n(?:[ \t]*\n)*[ \t]+)?`

var (
	pemRE = regexp.MustCompile(`(?s)-----BEGIN (?:[A-Z0-9]+ )*PRIVATE KEY(?: BLOCK)?-----.*?` +
		`(?:-----END (?:[A-Z0-9]+ )*PRIVATE KEY(?: BLOCK)?-----|\z)`)
	// Python's Unicode \b becomes a captured boundary character that the replacement re-emits.
	keyRE  = regexp.MustCompile(`(^|[^A-Za-z0-9])sk-[A-Za-z0-9_-]{16,}`) // vendor API keys by shape, after any non-alphanumeric, an underscore included
	authRE = regexp.MustCompile(`(?i)(^|[^\pL\pN_])((?:Bearer|Basic)[ \t]+[A-Za-z0-9._~+/=-]+)`)
	urlRE  = regexp.MustCompile(`(?i)(^|[^a-z0-9+.-])([a-z][a-z0-9+.-]*://[^<>"'` + pytext.SpaceBody + `]+)`)
	// regexp2: the lookbehind must see the char before the search start; three lookaheads.
	namedRE = regexp2.MustCompile(`(?is)((?<![\w-])(?=[\w-]*(?:token|password|passwd|secret|api[_-]?key|authorization))`+
		`[\w-]+["']?[ \t]*[:=]`+gap+`)(?:[!&][^\s]*`+gap+`){0,2}(`+
		`"""(?:\\.|(?!""").)*(?:"""|\z)|'''(?:\\.|(?!''').)*(?:'''|\z)|`+
		`"(?:\\.|[^"\\])*(?:"|\z)|'(?:''|\\.|[^'\\])*(?:'|\z)|[^\n]+)`, regexp2.None)
)

// Redact is privacy.redact: best-effort credential removal before an outbound LLM request. Every
// replacement re-emits the newlines it consumed, so line numbers survive redaction.
func Redact(text string) string {
	text = pemRE.ReplaceAllStringFunc(text, func(m string) string {
		return "[REDACTED]" + strings.Repeat("\n", strings.Count(m, "\n"))
	})
	text = supplychain.ReplaceSecrets(text, func(string) string { return "[REDACTED]" })
	text = authRE.ReplaceAllString(text, "${1}[REDACTED]")
	text = keyRE.ReplaceAllString(text, "${1}[REDACTED]")
	text = redactNamed(text)
	var b strings.Builder
	last := 0
	for _, m := range urlRE.FindAllStringSubmatchIndex(text, -1) {
		b.WriteString(text[last:m[4]])
		b.WriteString(supplychain.SanitizeSource(text[m[4]:m[5]]))
		last = m[1]
	}
	b.WriteString(text[last:])
	return b.String()
}

// redactNamed is the _NAMED loop over code points: a credential-shaped key keeps its key,
// separator and gap; the value (with any YAML continuation lines deeper than the key's indent)
// becomes [REDACTED]. A sibling key on a later line is never read as the value.
// printable keeps the model's free text out of the terminal's and the viewer's control planes: every
// control character and every format character, the bidi overrides included, becomes a space.
func printable(s string) string {
	return strings.Map(func(r rune) rune {
		if unicode.IsControl(r) || unicode.Is(unicode.Cf, r) {
			return ' '
		}
		return r
	}, s)
}

func redactNamed(text string) string {
	runes := []rune(text)
	var b strings.Builder
	pos := 0
	for {
		m, _ := namedRE.FindRunesMatchStartingAt(runes, pos)
		if m == nil {
			break
		}
		start, end := m.Index, m.Index+m.Length
		key, value := m.Groups()[1], m.Groups()[2]
		prefix := strings.TrimRight(string(runes[lineStart(runes, start):start]), "\"'")
		indent := utf8.RuneCountInString(prefix) - utf8.RuneCountInString(pytext.LStrip(prefix))
		if strings.Trim(prefix, " \t-") == "" {
			indent = utf8.RuneCountInString(prefix)
		}
		if slices.Contains(runes[start:value.Index], '\n') && value.Index-lineStart(runes, value.Index) <= indent {
			b.WriteString(string(runes[pos:value.Index]))
			pos = value.Index
			continue
		}
		if r := runes[value.Index]; r != '"' && r != '\'' {
			for end < len(runes) && runes[end] == '\n' { // _CONTINUATION.match(text, end)
				ws := end + 1
				for ws < len(runes) && (runes[ws] == ' ' || runes[ws] == '\t') {
					ws++
				}
				rest := ws
				for rest < len(runes) && runes[rest] != '\n' {
					rest++
				}
				if rest > ws && ws-(end+1) <= indent {
					break
				}
				end = rest
			}
		}
		b.WriteString(string(runes[pos:start]))
		b.WriteString(key.String())
		b.WriteString("[REDACTED]")
		b.WriteString(strings.Repeat("\n", strings.Count(string(runes[key.Index+key.Length:end]), "\n")))
		pos = end
	}
	b.WriteString(string(runes[pos:]))
	return b.String()
}

// lineStart is text.rfind("\n", 0, i) + 1.
func lineStart(runes []rune, i int) int {
	for i > 0 && runes[i-1] != '\n' {
		i--
	}
	return i
}
